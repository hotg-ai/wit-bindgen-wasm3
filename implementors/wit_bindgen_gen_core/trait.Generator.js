(function() {var implementors = {};
implementors["wit_bindgen_gen_wasm3"] = [{"text":"impl <a class=\"trait\" href=\"wit_bindgen_gen_core/trait.Generator.html\" title=\"trait wit_bindgen_gen_core::Generator\">Generator</a> for <a class=\"struct\" href=\"wit_bindgen_gen_wasm3/struct.Wasmtime.html\" title=\"struct wit_bindgen_gen_wasm3::Wasmtime\">Wasmtime</a>","synthetic":false,"types":["wit_bindgen_gen_wasm3::Wasmtime"]}];
if (window.register_implementors) {window.register_implementors(implementors);} else {window.pending_implementors = implementors;}})()