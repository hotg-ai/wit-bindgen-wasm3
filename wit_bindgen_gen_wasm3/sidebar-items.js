initSidebarItems({"enum":[["Async",""]],"struct":[["Opts",""],["Wasmtime",""]]});