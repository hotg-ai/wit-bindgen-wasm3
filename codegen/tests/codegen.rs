#[macro_use]
#[cfg(test)]
extern crate pretty_assertions;

use std::collections::BTreeMap;

use wit_bindgen_gen_core::{wit_parser::Interface, Direction, Files, Generator};
use wit_bindgen_gen_wasm3::Opts;

fn generate(filename: &str, src: &str, direction: Direction) -> String {
    let interface = Interface::parse(filename, src).unwrap();
    let mut files = Files::default();

    let mut imports = Vec::new();
    let mut exports = Vec::new();

    match direction {
        Direction::Import => imports.push(interface),
        Direction::Export => exports.push(interface),
    }

    Opts {
        rustfmt: true,
        ..Default::default()
    }
    .build()
    .generate_all(&imports, &exports, &mut files);

    let files: BTreeMap<_, _> = files.iter().collect();
    assert_eq!(files.len(), 1);
    std::str::from_utf8(&files["bindings.rs"])
        .unwrap()
        .to_string()
}

macro_rules! integration_tests {
    ($( $name:ident : $filename:literal ),* $(,)?) => {
        mod import {
            use super::*;

            $(
                #[test]
                fn $name() {
                    let src = include_str!($filename);
                    let $name = generate($filename, src, Direction::Import);

                    insta::assert_snapshot!($name);
                }
            )*
        }

        mod export {
            use super::*;

            $(
                #[test]
                fn $name() {
                    let src = include_str!($filename);
                    let $name = generate($filename, src, Direction::Export);

                    insta::assert_snapshot!($name);
                }
            )*
        }
    };
}

integration_tests! {
    single_function: "thunk.wit",
    empty: "empty.wit",
    record: "custom-record.wit",
}
