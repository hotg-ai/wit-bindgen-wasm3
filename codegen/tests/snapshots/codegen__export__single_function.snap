---
source: codegen/tests/codegen.rs
expression: single_function
---
pub mod thunk {
    #[allow(unused_imports)]
    use wit_bindgen_wasm3::rt::{anyhow, wasm3};
    pub trait Thunk: Sized {
        fn start(&mut self);
    }

    pub fn register<U>(
        module: &wit_bindgen_wasm3::rt::wasm3::Module<'_>,
        state: U,
    ) -> wit_bindgen_wasm3::rt::wasm3::Result<()>
    where
        U: Thunk + Send + Sync + 'static,
    {
        let state = Arc::new(Mutex::new(state));
        {
            let state = Arc::clone(&state);
            module.link_closure(
                "thunk",
                "start",
                move |mut caller: wit_bindgen_wasm3::rt::wasm3::CallContext<'_>| {
                    let state = state.lock().expect("lock was poisoned");
                    state.start();
                    Ok(())
                },
            )?;
        }
        Ok(())
    }
}

